Add groupSortedArray() function

palegre-tiny opened the pull request:
Changelog category (leave one):

New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New aggregation function groupSortedArray to obtain an array of first N values.

palegre-tiny committed:
Add groupSortedArray() function

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=34055) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=34055) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=34055) it.</sub>

palegre-tiny commented:
Detailed:
Added a new aggregation function:

Usage 1:
**groupSortedArray**(n)(param1)

It returns an array with the n first values from a field param1 sorted by itself.
```
 select groupSortedArray(2)(number) from numbers(100)
┌─groupSortedArray(2)(number)─┐
│ [0,1]                       │
└─────────────────────────────┘
```

Usage 2:
**groupSortedArray**(n)(param1, param2)

It returns an array with the n first values from a field param1 sorted param2 (field or expression).

SELECT groupSortedArray(5)(number, number) FROM numbers(1000000)

```
SELECT groupSortedArray(5)(number, -number) FROM numbers(100)
┌─groupSortedArray(5)(number, negate(number))─┐
│ [99,98,97,96,95]                            │
└─────────────────────────────────────────────┘
```

UnamedRus commented:
What if we want to sort by multiple columns?

```
SELECT groupSortedArray(5)(number, -number, number % 10) FROM numbers(100)

SELECT groupSortedArray(5)(number, (-number, number % 10)) FROM numbers(100)
```

I like first approach more, because sorting of tuple much slower.

palegre-tiny commented:
> 

It is not supported. Initially it only accepts integer expressions for this second parameter.

evillique commented on docs/en/sql-reference/aggregate-functions/reference/groupsortedarray.md:
```diff
@@ -0,0 +1,48 @@
+
+# groupSortedArray {#groupSortedArray}
+
+Returns an array with the first N items in ascending order.
```
Should the original column be sorted? Does it include duplicates in the final array?

palegre-tiny replied:
No, the original column can be sorted or not. 
Yes, it includes duplicates in the final array.

evillique commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,229 @@
+        if (cur < threshold)
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
```
It looks to me like an insertion sorting algorithm, or am I missing something?

palegre-tiny replied:
Yes, I need to get the first N elements. I made tests with other functions like std::nth_element() but I found this code more efficient.

evillique replied:
It would seem to me that this is only true up to a certain size, have you tested your implementation against std::nth_element and such (std::sort, std::partial_sort, std::partial_sort_copy, maybe even use std::priority_queue) for different Ns?

Because if it's true then we can make `if (x < threshold < y) { *use other implementation* }` and make everything even faster :)

palegre-tiny replied:
I created a repo to measure different methods. Maybe I missed something but I didn't find anything faster.

https://github.com/palegre-tiny/testfirstelements

palegre-tiny replied:
It was not my intention to code a new sorting function but I found it is much faster than other standard alternatives:

Current test results:
map 	  312132ns
sort   	1009483ns
custom1       19411ns
nth             146247ns

palegre-tiny replied:
I think I finally understood it, thanks.
I have added another selection algorimth for big array sizes, current code is 10x faster for short array sizes but for long sizes it is much more slower than std::nth_element (up to 100x). So I am using both, mine for threshold < 1000 and the standard one for bigger cases. 

I have also added performance tests for all (short, long) array sizes.

evillique commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,229 @@
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
+                results[z] = results[z - 1];
```
Also, this seems to cause heap-buffer-overflow in [some](https://s3.amazonaws.com/clickhouse-test-reports/34055/451d0c09bf158bf2e6a4239781ca2ae261eb192d/fuzzer_astfuzzerasan,actions//report.html) cases.

palegre-tiny replied:
Thanks, I am fixing it.

palegre-tiny commented:
> > What if we want to sort by multiple columns?
> > ```
> > SELECT groupSortedArray(5)(number, -number, number % 10) FROM numbers(100)
> > 
> > SELECT groupSortedArray(5)(number, (-number, number % 10)) FROM numbers(100)
> > ```
> > 
> >   
> > I like first approach more, because sorting of tuple much slower.
> 
> It is not supported. Initially it only accepts integer expressions for this second parameter.

I'm preparing a fixup supporting mode 2. Sorting any field/tuple sorting.

palegre-tiny commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,229 @@
+            TIndex values[batch_size];
+            memcpy(values, ref.data, batch_size * sizeof(TIndex));
+            size_t num_results = std::min(this->threshold, batch_size);
+            size_t * bestRows = new size_t[batch_size];
```
Ok, I will remove this new/delete calls.

palegre-tiny committed:
fixup! Add groupSortedArray() function

- Fix memory access
- Support any type as sorting parameter
- Fix tests
- Rewrite/simplify function addBatchSinglePlace

alexey-milovidov commented:
If the function named `groupSortedArray`, it may sound like 
"group sorted items into array" - incorrect
"group items into sorted array" - correct

If the function named `groupArraySorted`, it will sound like 
"group items into array and get it sorted" - correct

palegre-tiny committed:
fixup! Add groupSortedArray() function

filimonov commented:
See also #3708

filimonov commented on src/AggregateFunctions/AggregateFunctionGroupSortedArray.h:
```diff
@@ -0,0 +1,244 @@
+        if (cur < threshold)
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z > cur; z--)
```
Thouse embedded loops sound like O(n*m), where m is elements we need and n is number of elements in col. If m is high - it's square time. I guess we can do better

palegre-tiny replied:
Yes, honestly speaking I was not thinking in high array sizes, I will think about it.

palegre-tiny replied:
Totally right, performance is very poor for large array sizes, I am providing two different algoritmh for short/long arrays.

filimonov commented on docs/en/sql-reference/aggregate-functions/reference/groupsortedarray.md:
````diff
@@ -0,0 +1,48 @@
+Or the last 10:
+
+``` sql
+SELECT groupSortedArray(10)(number, -number) FROM numbers(100)
````
Alternative is storing lambda in the type (may complicate things)

palegre-tiny replied:
Yes, both parameters are typically not related so we simply store both in the serialization.

filimonov commented on src/AggregateFunctions/AggregateFunctionGroupSortedArrayData.h:
```diff
@@ -0,0 +1,162 @@
+
+    void merge(const AggregateFunctionGroupSortedArrayDataBase & other)
+    {
+        values.insert(other.values.begin(), other.values.end());
```
Both sides are sorted. Why not merge sort?  Also with limit

palegre-tiny replied:
Thanks, I will try to do like that.

palegre-tiny replied:
Do you mean using map::merge / set::merge...? I'm not sure I can use them here.

palegre-tiny commented on docs/en/sql-reference/aggregate-functions/reference/groupsortedarray.md:
```diff
@@ -0,0 +1,48 @@
+
+-   `N` – The number of elements to return.
+
+If the parameter is omitted, default value 10 is used.
```
Thanks, I will check how groupArray does.

palegre-tiny replied:
I think this function is not supposed to generate so large arrays, the advantage of using groupArraySorted() is to not sort all the elements but to select the top ones. For this purpose I would just ORDER all the registers in the query and take them with groupArray().

palegre-tiny committed:
fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! Add groupSortedArray() function

fixup! Add groupSortedArray() function

fixup! Add groupSortedArray() function

palegre-tiny committed:
fixup! Evolve to TinySum

palegre-tiny committed:
fixup! Evolve to TinySum

palegre-tiny commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

palegre-tiny committed:
fixup! Add groupSortedArray() function

palegre-tiny commented:
Thank you for your inputs, I have been adding the requested modifications and I think it is again ready for review.

List of modifications:
- Renamed to groupArraySorted
- New selection function for long arrays size (More than 1000)
- Added performance test for long arrays size
- Fixed memory access issue
- Simplified aggregation class creation methods
- Fixed syntax checks and other CI issues
- Added support for any type in the second parameter (sorted by)
- Rewritten addBatchSinglePlace() function to remove dynamic allocations and to make the code mode understandable

I can still see a failing check but I think it is not related with my code. (test_s3_zero_copy_replication/test.py::test_s3_zero_copy_drop_detached - FAILED)

jkuklis commented:
Hey!

I would like to ask: @palegre-tiny and PR reviewers, do you know whether it would be difficult to add support for:
* `groupArraySorted` `-Array` extensions, like `groupArraySortedArray` and `groupUniqArraySortedArray`,
* `SimpleAggregateFunction(groupArraySorted)` and with `-Array` extensions?

For context, we will likely need these features for a project that we discuss currently at our company. We were willing to implement these functionalities from scratch on our own, then it was good news for us that somebody else is working on a similar design, so that we have a better starting point and a source of inspiration.

We would like to use `SimpleAggregateFunction(groupArraySortedArray)` and `SimpleAggregateFunction(groupUniqArraySortedArray)` in `AggregatingMergeTree`. Extensions with `-Array` flatten the input arrays into a single output array, and `SimpleAggregateFunction` is a wrapper for aggregate functions like `groupArrayArray`.

Could you share some thoughts about these extensions above:
* Can we reuse the design for `groupArraySorted` for `-Array` extensions? Are there any obstacles in the current design?
* Can we simply enable `groupArraySorted` as a valid choice for `SimpleAggregateFunction` in `checkSupportedFunctions` in src/DataTypes/DataTypeCustomSimpleAggregateFunction.cpp?

There are likely many details to work out here anyway, but if you have enough context to give us hints, it would be great!


Thanks!

Jakub Kuklis
Contentsquare

palegre-tiny committed:
fixup! fixup! Add groupSortedArray() function

palegre-tiny commented:
> 
Thanks Jakub for your comments.

1. About groupArraySorted +Array extensions

If I understood correctly you suggest something like this:

```sql
table_test_array
┌─a───────────┬─b─┐
│ [1,3,5,7,9] │ 2 │
│ [0,2,4,6,8] │ 1 │
└─────────────┴───┘

groupArraySortedArray(a) from table_test_array
┌─groupArraySortedArray(a)─┐
│ [0,1,2,3,4]              │
└──────────────────────────┘
```
And that is clear, but what if we want to use second parameter (sort expression), we would do the following:
```sql
groupArraySortedArray(a,b) from table_test_array
┌─groupArraySortedArray(a,b)─┐
│ [0,2,4,6,8]                │
└────────────────────────────┘
```
this case we would not split the arrays because each has different weight.

I don't see many problems in the implementation, extracting each value from the array and processing separately. I have more doubts about if it is clear from ussage perspective.



2. Use with SimpleAggregateFunction

I don't think we can use it, again when using 2 parameters we need to store the weight (sort expression) attached to the values, even if at the end we only show the values.

palegre-tiny commented:
https://github.com/Mergifyio update

palegre-tiny commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

palegre-tiny commented:
For sorted by second parameter usage  + array extension: It is necessary to use array in both parameters:
```sql
select groupArraySortedArray([1, 2, 3], [10, 0, 5])

┌─groupArraySortedArray([1, 2, 3], [10, 0, 5])─┐
│ [2,3,1]                                      │
└──────────────────────────────────────────────┘
```

palegre-tiny commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

palegre-tiny commented:
Hi @filimonov @evillique,
It's been a while since the last review/comment, is there anything I can do to continue this?
Thank you

palegre-tiny commented:
https://github.com/Mergifyio update

palegre-tiny commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into groupSortedArray

alexey-milovidov commented on src/AggregateFunctions/AggregateFunctionGroupArraySorted.h:
```diff
@@ -0,0 +1,310 @@
+
+template <typename TColumn, typename TFilter = void>
+size_t
+getFirstNElements_low_threshold(const TColumn * data, int num_elements, int threshold, size_t * results, const TFilter * filter = nullptr)
```
Wrong style.

alexey-milovidov commented on src/AggregateFunctions/AggregateFunctionGroupArraySorted.h:
```diff
@@ -0,0 +1,310 @@
+                continue;
+        }
+
+        //Starting from the highest values and we look for the immediately lower than the given one
```
@evillique Wrong style.

alexey-milovidov commented:
@evillique I did not understand the algorithm of this function in five seconds. Let's move it under "experimental" flag.
